server/src/main/java/org/elasticsearch/search/SearchHit.java

```diff
@@ -482,6 +483,42 @@ public void setDocumentField(String fieldName, DocumentField field) {
         }
     }
 
+    /**
+     * Whether this search hit has any lookup fields
+     */
+    public boolean hasLookupFields() {
+        return getDocumentFields().values().stream().anyMatch(doc -> doc.getLookupFields().isEmpty() == false);
+    }
+
+    /**
+     * Resolve the lookup fields with the given results and merge them as regular fetch fields.
+     */
+    public void resolveLookupFields(Map<LookupField, List<Object>> lookupResults) {
+        if (lookupResults.isEmpty()) {
+            return;
+        }
+        final List<String> fields = new ArrayList<>(documentFields.keySet());
+        for (String field : fields) {
+            documentFields.computeIfPresent(field, (k, docField) -> {
+                if (docField.getLookupFields().isEmpty()) {
```

Comment:
Maybe we don't want to touch a field unless it's a lookup field? I'm wondering a bit about performance here because it's common for hits to have a bunch of fields (Kibana Discover requests all fields `*` by default).

Reply:
I think the performance is the same, see [HashMap#computeIfPresent](https://github.com/openjdk/jdk/blob/0af356bb4bfee99223d4bd4f8b0001c5f362c150/src/java.base/share/classes/java/util/HashMap.java#L1253).